File: kt/src/main/kotlin/dev/pragmastat/distributions/Uniform.kt
package dev.pragmastat.distributions

import dev.pragmastat.Rng

/**
 * Uniform distribution on [min, max).
 *
 * All values within the interval have equal probability.
 *
 * @property min Lower bound (inclusive).
 * @property max Upper bound (exclusive).
 * @throws IllegalArgumentException If min >= max.
 */
class Uniform(private val min: Double, private val max: Double) : Distribution {
    init {
        require(min < max) { "min must be less than max" }
    }

    override fun sample(rng: Rng): Double {
        return min + rng.uniform() * (max - min)
    }
}

File: ts/src/distributions/uniform.ts
/**
 * Uniform distribution on a bounded interval.
 */

import { Rng } from '../rng';
import { Distribution } from './distribution';

/**
 * Uniform distribution on [min, max).
 *
 * @example
 * const rng = new Rng("demo-dist-uniform");
 * const dist = new Uniform(0, 10);
 * const sample = dist.sample(rng);
 */
export class Uniform implements Distribution {
  private readonly min: number;
  private readonly max: number;

  /**
   * Create a new uniform distribution on [min, max).
   *
   * @param min - Lower bound (inclusive).
   * @param max - Upper bound (exclusive).
   * @throws Error if min >= max.
   */
  constructor(min: number, max: number) {
    if (min >= max) {
      throw new Error('min must be less than max');
    }
    this.min = min;
    this.max = max;
  }

  sample(rng: Rng): number {
    return this.min + rng.uniform() * (this.max - this.min);
  }

  samples(rng: Rng, count: number): number[] {
    return Array.from({ length: count }, () => this.sample(rng));
  }
}
